drivers/ieee802154/ieee802154_dw1000.c

```diff
@@ -769,11 +769,137 @@ static int dwt_filter(const struct device *dev,
 	return -ENOTSUP;
 }
 
-static int dwt_set_power(const struct device *dev, int16_t dbm)
+uint8_t dwt_encode_power_tx(int power)
 {
-	struct dwt_context *ctx = dev->data;
+	uint8_t coarse = DWT_TX_POWER_CTRL_OFF;
+	uint8_t fineTuneVal = 0;
+
+	if (power >= 0) {
+		coarse--;
+		while (power >= DWT_TX_POWER_CTRL_COARSE_STEPS && coarse > 0) {
+			power -= DWT_TX_POWER_CTRL_COARSE_STEPS;
+			coarse--;
+		}
+		fineTuneVal = (uint8_t)(power / DWT_TX_POWER_CTRL_FINE_STEPS + 0.5) &
+			      DWT_TX_POWER_CTRL_FINE_MASK;
+	}
+	uint8_t result = (coarse << 5) | fineTuneVal;
+	return result;
+}
+
+float dwt_decode_power_tx(uint8_t tx_power_control)
+{
+	uint8_t coarseSetting = (tx_power_control & DWT_TX_POWER_CTRL_COARSE_MASK) >> 5;
+	if (coarseSetting == 0x07) {
+		return 0;
+	}
+	uint8_t fineSetting = tx_power_control & DWT_TX_POWER_CTRL_FINE_MASK;
+	float coarseGain = 15.0 - (coarseSetting * 2.5);
+	float fineGain = fineSetting * 0.5;
+	float totalGain = coarseGain + fineGain;
+
+	return totalGain;
+}
+
+static int dwt_set_txpower(const struct device *dev, int16_t dbm)
+{
+	uint32_t tx_power = dwt_reg_read_u32(dev, DWT_TX_POWER_ID, 0);
+	uint32_t sys_cfg = dwt_reg_read_u32(dev, DWT_SYS_CFG_ID, 0);
+
+	if (sys_cfg & DWT_SYS_CFG_DIS_STXP) {
+		/* Manual Transmit Power Control */
+		float phr = dwt_decode_power_tx((tx_power >> DWT_TX_POWER_TXPOWPHR_SHIFT) & 0xFF);
+		float sd = dwt_decode_power_tx((tx_power >> DWT_TX_POWER_TXPOWSD_SHIFT) & 0xFF);
+		phr += dbm;
```

Comment:
Yes, this was always clear. The current TX power setting (AFAIK) could only have values from the table 19/20 since. Anyways the approach have changed and now the function reads directly the (decoded) values from the tables depending on the active channel